endpoint: Implement new Invalid endpoint state

christarazi opened the pull request:
See commit msgs.

Endpoint (EP) creation can fail for many reasons due to invalid data, such as IP conflict or if it's created with reserved labels. When EP creation failed, EPs can be stuck in the "waiting-for-identity" state, and left to be garbage collected. However, even though the object would be garbage collected, the metric representing which state the EP is in, is never decremented, thus a leak. 

This PR fixes this by introducing a new EP state representing an "invalid" EP. When EP creations fails, the EP will transition from its current state to "invalid", thus decrementing the current state metric.

```release-note
Fix leaking endpoint state metric
```

christarazi committed:
endpoint: Improve tests to assert against metrics

This commits augments the endpoint state transition unit tests to assert
against incrementing / decrementing metrics. It also adds new tests to
cover the new endpoint state, "invalid".

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
daemon: Add assertions for endpoint state metrics

This commit adds new assertions on endpoint state metrics to the
existing test cases. This commit helps validate that when an endpoint
fails during creation, it will be deleted, and the bookkeeping of the
endpoint state metrics is correct.

Additionally, this commits increases the code coverage around the
creation of endpoints when they fail.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
metrics: Add getter for gauge values

This is useful for observing gauge values in tests. Previously, only
counter values had a getter.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

aanm commented on pkg/metrics/metrics.go:
```diff
@@ -1147,6 +1147,16 @@ func GetCounterValue(m prometheus.Counter) float64 {
 	return 0
 }
 
+// GetGaugeValue returns the current value stored for the gauge
```
Write in the comment that this function is only used in tests.

christarazi replied:
Fixed

aanm commented on daemon/cmd/endpoint.go:
```diff
@@ -174,7 +174,15 @@ func (d *Daemon) fetchK8sLabelsAndAnnotations(nsName, podName string) (*slim_cor
+			NoIPRelease: true,
+		})
+	}
+
```
I would suggest to create another function instead of reusing this one.

aanm commented on daemon/cmd/endpoint.go:
```diff
@@ -214,17 +222,17 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
 	oldEp := d.endpointManager.LookupCiliumID(ep.ID)
 	if oldEp != nil {
-		return invalidDataError(ep, fmt.Errorf("endpoint ID %d already exists", ep.ID))
+		return d.invalidDataError(ep, true, fmt.Errorf("endpoint ID %d already exists", ep.ID))
```
this should be false, not true. If an endpoint already exists aren't we potentially deleting that same endpoint?

christarazi replied:
Good catch, will fix

christarazi replied:
On second thought, this would suffer from the same thing I described below. Not deleting this endpoint object would leak the metric. Are you saying that because in this specific case, both endpoint have the same ID, that would result in the original endpoint object being deleted?

christarazi replied:
Disregard my comments--was reading your comments from in order and didn't see your proposal at the end :+1:

aanm commented on daemon/cmd/endpoint.go:
```diff
@@ -214,17 +222,17 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
 	oldEp = d.endpointManager.LookupContainerID(ep.GetContainerID())
 	if oldEp != nil {
-		return invalidDataError(ep, fmt.Errorf("endpoint for container %s already exists", ep.GetContainerID()))
+		return d.invalidDataError(ep, true, fmt.Errorf("endpoint for container %s already exists", ep.GetContainerID()))
```
this should be false, not true. If an endpoint already exists aren't we potentially deleting that same endpoint?

christarazi replied:
Good catch, will fix

aanm commented on daemon/cmd/endpoint.go:
```diff
@@ -240,27 +248,27 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
 		oldEp, err := d.endpointManager.Lookup(id)
 		if err != nil {
-			return invalidDataError(ep, err)
+			return d.invalidDataError(ep, true, err)
```
this should be false, not true. same reasons above

christarazi replied:
This would also be a very similar situation that I described below.

aanm commented on daemon/cmd/endpoint.go:
```diff
@@ -240,27 +248,27 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
+			return d.invalidDataError(ep, true, err)
 		} else if oldEp != nil {
-			return invalidDataError(ep, fmt.Errorf("IP %s is already in use", id))
+			return d.invalidDataError(ep, true, fmt.Errorf("IP %s is already in use", id))
```
this should be false, not true. same reasons above

christarazi replied:
This was the exact error case that a customer ran into. This specific case was causing the errant metric.

aanm commented on daemon/cmd/endpoint.go:
```diff
@@ -240,27 +248,27 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
 
 	if err = endpoint.APICanModify(ep); err != nil {
-		return invalidDataError(ep, err)
+		return d.invalidDataError(ep, true, err)
```
this should be false, not true. same reasons above

christarazi replied:
Why would this be problematic? I'm not clear on what this check is validating.

aanm commented on daemon/cmd/endpoint.go:
```diff
@@ -174,7 +174,15 @@ func (d *Daemon) fetchK8sLabelsAndAnnotations(nsName, podName string) (*slim_cor
 	return p, containerPorts, identityLabels, infoLabels, annotations, nil
```
In this file, instead of the proposed changes I suggest to do the following (on all error conditions):
```
		ep.SetState(endpoint.StateInvalidEndpoint, "Invalid endpoint")
```

and in `pkg/endpoint/endpoint.go`
```patch
diff --git a/pkg/endpoint/endpoint.go b/pkg/endpoint/endpoint.go
index 8bd438963..bd504c97a 100644
--- a/pkg/endpoint/endpoint.go
+++ b/pkg/endpoint/endpoint.go
@@ -96,6 +96,8 @@ const (
        // StateRestoring is used to set the endpoint is being restored.
        StateRestoring = string(models.EndpointStateRestoring)
 
+       // TODO Needs an new state in the api/v1/models`
+       StateInvalidEndpoint = string(models.StateInvalidEndpoint)
+
        // IpvlanMapName specifies the tail call map for EP on egress used with ipvlan.
        IpvlanMapName = "cilium_lxc_ipve_"
 )
@@ -1319,7 +1321,7 @@ func (e *Endpoint) setState(toState, reason string) bool {
                }
        case StateWaitingForIdentity:
                switch toState {
-               case StateReady, StateDisconnecting:
+               case StateReady, StateDisconnecting, StateInvalidEndpoint:
                        goto OKState
                }
        case StateReady:
@@ -1389,7 +1391,7 @@ OKState:
 
        // Since StateDisconnected is the final state, after which the
        // endpoint is gone, we should not increment metrics for this state.
-       if toState != "" && toState != StateDisconnected {
+       if toState != "" && (toState != StateDisconnected || toState != StateInvalidEndpoint) {
                metrics.EndpointStateCount.
                        WithLabelValues(toState).Inc()
        }
@@ -1403,7 +1405,7 @@ func (e *Endpoint) BuilderSetStateLocked(toState, reason string) bool {
        // Validate the state transition.
        fromState := e.state
        switch fromState { // From state
-       case StateWaitingForIdentity, StateReady, StateDisconnecting, StateDisconnected:
+       case StateWaitingForIdentity, StateReady, StateDisconnecting, StateDisconnected, StateInvalidEndpoint:
                // No valid transitions for the builder
        case StateWaitingToRegenerate, StateRestoring:
                switch toState {
@@ -1456,7 +1458,7 @@ OKState:
 
        // Since StateDisconnected is the final state, after which the
        // endpoint is gone, we should not increment metrics for this state.
-       if toState != "" && toState != StateDisconnected {
+       if toState != "" && (toState != StateDisconnected || toState != StateInvalidEndpoint) {
                metrics.EndpointStateCount.
                        WithLabelValues(toState).Inc()
        }
```

christarazi committed:
endpoint: Add Invalid state

This commit adds a new Endpoint state called "invalid-endpoint". This
state represents endpoints that failed during their creation due to
invalid data. Previously, endpoints that failed during their creation
due to invalid data were ignored and left running. This would cause
leaking metrics as they'd likely be stuck in "waiting-for-identity"
which can be alarming for the user.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

joestringer commented on pkg/endpoint/endpoint_test.go:
```diff
@@ -244,115 +261,160 @@ func (s *EndpointSuite) TestEndpointState(c *C) {
+	}
+}
+
+func isFinalState(state string) bool {
+	return (state == StateDisconnected || state == StateInvalid)
+}
```
FWIW I think a function like this would be nice to have in the actual code as well, to reference from the core state transition function. But I don't think it matters enough to further change this before merging.

joestringer commented on daemon/cmd/endpoint.go:
```diff
@@ -176,6 +176,7 @@ func (d *Daemon) fetchK8sLabelsAndAnnotations(nsName, podName string) (*slim_cor
 
 func invalidDataError(ep *endpoint.Endpoint, err error) (*endpoint.Endpoint, int, error) {
 	ep.Logger(daemonSubsys).WithError(err).Warning("Creation of endpoint failed due to invalid data")
+	ep.SetState(endpoint.StateInvalid, "Invalid endpoint")
```
I checked through as well to see that we don't need this anywhere else such as `errorDuringCreation()`, but that path is fine because it invokes `deleteEndpointQuiet()` which would transition the endpoint into `StateDisconnecting`.

joestringer reviewed (approved):
Endpoint creation strikes again.
